Visible groups in jump screen

cooperka opened the pull request:
Makes labeled groups "visible" so they show up on the jump screen. You can then click on the group to see its children.

Currently, groups are invisible to the user, but with this new appearance, they become a selectable item in the jump screen. All children of the group are then hidden from the main screen, and will be shown on the inner screen instead, effectively giving these **visible** groups the exact same navigation behavior as existing **repeatable** groups.

To make a group visible, simply make sure it has a `ref` and a `label` like so:

```xml
<group ref="...">
    <label>Some label</label>
    ...
</group>
```

In more technical terms, groups must be both **presentable** and **logical** (see [docs](https://opendatakit.github.io/xforms-spec/#groups) for definitions) in order to be visible.

Example with old, existing UI:

<div>
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51065358-672f9d80-15d2-11e9-830d-53503c080625.png">
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51065359-672f9d80-15d2-11e9-96ac-e47d7dbef52a.png">
</div>

<br/>

Example with new, grouped UI:

<div>
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51065772-9e9f4980-15d4-11e9-9641-7bfe336083b9.png">
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51065773-9e9f4980-15d4-11e9-93eb-b1b9bd2cfbc8.png">
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51065774-9e9f4980-15d4-11e9-873f-dd8ceed9ba7a.png">
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51065775-9e9f4980-15d4-11e9-8710-54f0e2e1b9b2.png">
</div>

#### What has been done to verify that this works as intended?

Extensive manual testing with:
- visible_groups ([XML](https://drive.google.com/open?id=1JabZFs4NdFrLSjkqdDtXc2EnJWuK_IoD) / [XLS](https://drive.google.com/open?id=1DwXUPtcNly8mPRPAIreIhq7ME_D3X0W7gllCmSAX2g8)) (grouped widgets; the "Date and time widgets" group is marked as visible, and has a nested "Inner group" that's also visible)
- tcc-demo ([XML](https://github.com/opendatakit/collect/files/2781332/tcc-demo.xml.txt))

Navigating in and out of nested groups, using the back button, jumping back and forth between the FormEditor and FormHierarchy views.

#### Why is this the best possible solution? Were any other approaches considered?

Allowing groups to be visible gives more power to form designers who may want to prevent certain groups from cluttering the jump screen. It also makes it easier to focus on the questions within those groups if desired.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Labeled groups will require one more tap to navigate into, and the children of the group will be hidden from the previous page.

Regression risks include any code that previously assumed that regular groups could never be "screen events". There were several methods that checked for **repeatable** groups, which now also need to check for **visible** groups as well. All of the logic I could find has been updated accordingly.

#### Do we need any specific form for testing your changes? If so, please attach one.

Linked above.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

Yes, documenting the new logic [here](http://xlsform.org/en/#appearance). Issue [here](https://github.com/opendatakit/docs/issues/931).

#### Known issues being addressed
- [x] Groups with dynamic labels (e.g. a `ref`) aren't detected as "visible"
- [x] In some forms, the last item inside the group is shown outside the group
- [x] In forms without group `ref`s, other groups outside the group seem to be shown inside the group
- [ ] Update the docs issue above for the new default behavior

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

cooperka committed:
De-duplicate idential groupIsFieldList functions

cooperka committed:
Add getAppearanceAttr for form index

cooperka committed:
Add braces to big switch statement to prevent var collisions

cooperka committed:
Display groups that have 'nested' appearance attribute

Doesn't handle taps yet.

cooperka committed:
Show children of displayable groups in their own screen

cooperka committed:
Change key from 'nested' to 'visible'; make static

cooperka committed:
Fix stepping logic now that groups can be visible

Now a 'screen event' entails either a repeat or a visible group.

cooperka committed:
Update appearance to hierarchy-visible

cooperka committed:
Fix updated HierarchyElement type VISIBLE_GROUP

cooperka committed:
Fix a few PMD rule violations

cooperka commented:
Tests are passing locally but CI is repeatedly failing with `Could not GET 'https://jcenter.bintray.com/net/java/dev/javacc/javacc/5.0/javacc-5.0.jar'. Received status code 504 from server: Gateway Time-out`

~~If an admin can re-run the build later on, hopefully that's resolved.~~ Thanks!

yanokwa commented:
@cooperka This is the first I've heard of this appearance. We should first discuss this appearance on the development category of the forum where others are likely to see it.

cooperka commented:
Sounds good @yanokwa, I just posted it [here](https://forum.opendatakit.org/t/new-xform-appearance-hierarchy-visible-groups/16858).

grzesiek2010 commented:
@cooperka 

Please let me know once you finish work on this. I saw you agree on the approach without `appearance="hierarchy-visible"`

Then I can review.

cooperka committed:
Check for label instead of visibility attribute

cooperka committed:
Remove redundant forceStepOverGroup method

cooperka committed:
Don't skip over the next group when rendering hierarchy

Fix for when two visible groups are in a row.
Would previously miss the second one.

cooperka committed:
Add 'folder' icon to visible groups

cooperka committed:
Use better Material Design padding on hierarchy icons

cooperka commented:
@grzesiek2010 I added a few more commits and rebased on latest master. The last commit edf64d1 implements the default behavior we discussed on the forum; your review is now welcome.

I'm concerned that the current grouping behavior may be too aggressive. When entering a repeatable group, depending on the form structure, it may require a second click to actually get to the questions:

<div>
<img width="180" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51066047-18840280-15d6-11e9-9cca-613d9e62ab4a.png">
<img width="180" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51066048-191c9900-15d6-11e9-99d3-98bc0863fc01.png">
<img width="180" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51066049-191c9900-15d6-11e9-8a12-382ec4acb8e7.png">
</div>

<br/>

I'm not sure if this is a "feature" as a result of a poorly designed form, or an implementation bug. I'd appreciate your input here (or I'll have more time to diagnose by Tuesday). Form used: [XML](https://drive.google.com/open?id=1zOvDEZz6vwQr5IBT91k7MuDMZ-hJDats) / [XLS](https://drive.google.com/open?id=1xVv682pbBMQ_MTZdNnj6aP11jhPp6fT8WQ42oEr7WaA).

yanokwa commented:
@cooperka The odd behavior makes sense to me because repeats are a type of group. I think/hope most forms don't have a repeat followed by a group, but I bet it's not uncommon. In your specific case, I believe @smoyth and @lognaturel came up with this group within a repeat construction as a way to add dynamic labels. 

My bias is to keep it for now since it's technically correct. Then we can have a follow on discussion on if/how we can collapse those in someway.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -663,6 +682,23 @@ public int stepToOuterScreenEvent() {
+    /**
+     * Returns true if the group has a displayable label.
+     */
+    private boolean groupIsLabeled(FormIndex groupIndex) {
```
`isGroupLabeled` would be a better name here.

cooperka committed:
PR fix: Rename isGroupLabeled

mmarciniak90 commented:
@cooperka @yanokwa 
I noticed that something is incorect eg. in form:
[nested-repeats-complex.xml.txt](https://github.com/opendatakit/collect/files/2760442/nested-repeats-complex.xml.txt)

1. User adds 2 friends and 3 pets for each friend and one enemy
2. User opens jump screen (view 1)
3. User clicks on `Friends` and sees view 2
4. User clicks on `Fr1 (1)` and sees view 3. 

User is not able to open question where `Pet2` or `Pet3` were added because only `Pet1` is visible.
Opening `Pet1` group shows the question. When user clicked 2 times on `Go Up` he sees three groups of pets (view 5)

|version|view 1|view 2|view 3|view 4|  | view5|
|---|---|---|---|---|---|---|
|pr/2764|![screenshot_20190115-171111](https://user-images.githubusercontent.com/3907554/51193417-19c07280-18e9-11e9-9da9-65ad93256d0f.png)|![screenshot_20190115-171121](https://user-images.githubusercontent.com/3907554/51193483-42e10300-18e9-11e9-9789-ce5eebf61bad.png)|![screenshot_20190115-171134](https://user-images.githubusercontent.com/3907554/51193513-4f655b80-18e9-11e9-8b5f-696409dd3b68.png)|![screenshot_20190115-171147](https://user-images.githubusercontent.com/3907554/51193601-820f5400-18e9-11e9-8a2f-73e37b7f9794.png)|Go Up 2 times|![screenshot_20190115-171351](https://user-images.githubusercontent.com/3907554/51193667-b125c580-18e9-11e9-87b1-cddcf7519ea8.png)|
|1.18.2|![screenshot_2019-01-15-19-00-31](https://user-images.githubusercontent.com/3907554/51193851-22fe0f00-18ea-11e9-9fc5-9010706158ee.png)|![screenshot_2019-01-15-19-00-41](https://user-images.githubusercontent.com/3907554/51193867-285b5980-18ea-11e9-9f45-23de78ad296b.png)|![screenshot_2019-01-15-19-00-51](https://user-images.githubusercontent.com/3907554/51193877-2f826780-18ea-11e9-8bc1-0f9d80223570.png)|![screenshot_2019-01-15-19-11-51](https://user-images.githubusercontent.com/3907554/51193951-58a2f800-18ea-11e9-9ee1-f86c250b3e2a.png)|

mmarciniak90 commented:
@opendatakit-bot unlabel 'needs testing'

cooperka commented:
It looks like View 3 should either show items for Pet2 and Pet3 also, or it should show the parent Pets group instead. I'll look into what's going on, thanks!

cooperka committed:
Rename repeatGroupRef => visibleGroupRef

And move some comments around for clarity.

cooperka committed:
Save visible groups to visibleGroupRef, not just repeats

cooperka committed:
Reverse conditional to make logic more readable

cooperka committed:
Stop stringifying refs, just use TreeReference

cooperka committed:
Use stricter equals() to check repeat group membership

Just to be safe; this doesn't change anything for most forms.

cooperka committed:
Don't render other visible groups' instances

cooperka committed:
Prevent a redundant middle screen (common on some forms)

cooperka committed:
Remove redundant getLabel method

shortText falls back to longText automatically.

cooperka committed:
Use shortText to determine group label

Now labels with a `ref` will also be detected (previously only
explicit XML inner text for labels was supported).

cooperka committed:
Simplify logic for skipping past the group

Instead of going back to allow it to go forward,
simply `continue` the loop early.

This fixes a bug in some forms where the last item in a group is
displayed outside of the group.

cooperka commented:
Good news -- I was able to add some simple logic in d776315 to detect "label" groups and hide them from the navigation. Users will no longer have to click twice to get into a labeled repeat group, so the issue discussed above is no longer relevant.

It seems to be working well in every form I've been able to test with except for [tcc-demo](https://github.com/opendatakit/collect/files/2781332/tcc-demo.xml.txt) (XML). The issue is that all items in the form after the visible `Head of Household` group get nested inside the group when they're supposed to be outside. It's supposed to be:

```
Head of Household (group)
    [questions for Head of Household]
Household Members (repeat)
What number is your team?
Household Members 2 (repeat)
```

but instead it displays:

```
Head of Household (group)
    [questions for Head of Household]
    Household Members (repeat)
    What number is your team?
    Household Members 2 (repeat)
```

---

Update:

The reason for this is that the form doesn't use "logical groups" (valid according to the [docs](https://opendatakit.github.io/xforms-spec/#groups)). Thus the ref for the `Head of Household` group is simply `/data`, just like every other group in the form... whereas normally the ref for groups in other forms is set to something like `/data/head_of_household[1]`.

It seems to me that a `TreeReference` should still understand its hierarchy even if it hasn't explicitly been given a `ref`... shouldn't it? Relevant code is near [here](https://github.com/opendatakit/collect/pull/2764/files#diff-c4569759abab98c20df2aa91fe129bc8R392): `curGroup.isParentOf(currentRef, false)` returns `true` because every item has the same `TreeReference`.

@grzesiek2010 I'd appreciate your help if you have time to debug. Do you have a better idea for how we can detect group boundaries when everything has the same ref and TreeReference can't be used to determine lineage? Am I just confused? I'll keep debugging tomorrow [Tuesday] but I don't want to block the feature freeze.

cooperka commented:
@opendatakit-bot label "needs testing"

@mmarciniak90 I expect this to work well on all forms unless they don't have `ref` attributes on groups. Those have the known issue described above. If you find additional bugs please let me know.

cooperka committed:
Rename isGroupLabeled => isPresentationGroup for clarity

Matches language in the docs

cooperka committed:
Allow empty string for group name

Otherwise the hierarchy would be inconsistent depending on
text value.

cooperka committed:
Also check isLogicalGroup before making a group visible

This works, but it's gross. We should make JavaRosa support this explicitly.

cooperka commented:
The latest commit 6f600dc adds a check for "logical groups" and only makes a group visible if it's both **presentable** and **logical** (see [docs](https://opendatakit.github.io/xforms-spec/#groups) for definitions). This specific implementation is rather fragile, depending on assumptions about JavaRosa internals, but I plan to make a much simpler change in JavaRosa to support the same behavior more innately.

This means that groups **without** a `ref` will **not** show up in the hierarchy. Non-logical groups are hard to support here and are also seemingly somewhat rare (based on the forms I've personally come across).

Ready for full testing!

mmarciniak90 commented:
**Tested with success**

Verified on Android: 8.1, 7.0, 6.0, 5.1, 4.4, 4.2

Verified forms:
* All widgets
* fewRepeatsGroups
* nested-indexed-repeat
* nested-repeats
* nested-repeats-complex
* nRepeats
* TCC demo
* Visible Groups

| | |
|---|---|
|![screenshot_20190123-162227](https://user-images.githubusercontent.com/3907554/51616856-4b5dbd00-1f2b-11e9-9018-c4d7eea2a7d1.png)|![screenshot_20190123-154516](https://user-images.githubusercontent.com/3907554/51616857-4b5dbd00-1f2b-11e9-98a7-cbe749229ae0.png)|

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

cooperka commented:
Thanks for the merge @grzesiek2010!

Here's the follow-up PR with the JavaRosa simplification: https://github.com/opendatakit/collect/pull/2815.
